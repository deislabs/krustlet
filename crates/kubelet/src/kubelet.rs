///! This library contains code for running a kubelet. Use this to create a new
///! Kubelet with a specific handler (called a `Provider`)
use crate::config::Config;
use crate::node::{create_node, update_node};
use crate::server::start_webserver;
use crate::Provider;

use futures::{StreamExt, TryStreamExt};
use k8s_openapi::api::core::v1::Pod as KubePod;
use kube::{api::ListParams, runtime::Informer, Resource};
use log::{debug, warn};
use tokio::sync::Mutex;

use std::sync::Arc;

/// A Kubelet server backed by a given `Provider`.
///
/// A Kubelet is a special kind of server that handles Kubernetes requests
/// to schedule pods.
///
/// The Kubelet creates a listener on the Kubernetes API (called an Informer),
/// a webserver for API callbacks, and a periodic updater to let Kubernetes
/// know that the node is still running.
///
/// The Provider supplies all of the backend-specific logic. Krustlet will only
/// run one (instance of a) Provider. So a provider may be passed around from
/// thread to thread during the course of the Kubelet's lifetime.
pub struct Kubelet<P> {
    provider: Arc<Mutex<P>>,
    kube_config: kube::config::Configuration,
    config: Config,
}

impl<T: 'static + Provider + Sync + Send> Kubelet<T> {
    /// Create a new Kubelet with a provider, a kubernetes configuration,
    /// and a kubelet configuration
    pub fn new(provider: T, kube_config: kube::config::Configuration, config: Config) -> Self {
        Self {
            provider: Arc::new(Mutex::new(provider)),
            kube_config,
            config,
        }
    }

    /// Begin answering requests for the Kubelet.
    ///
    /// This will listen on the given address, and will also begin watching for Pod
    /// events, which it will handle.
    pub async fn start(&self) -> anyhow::Result<()> {
        let client = kube::Client::from(self.kube_config.clone());
        // Create the node. If it already exists, "adopt" the node definition
        create_node(&client, &self.config, T::ARCH).await;

        // Get the node name for use in the update loop
        let node_name = self.config.node_name.clone();
        // Start updating the node lease periodically
        let update_client = client.clone();
        let node_updater = tokio::task::spawn(async move {
            let sleep_interval = std::time::Duration::from_secs(10);
            loop {
                update_node(&update_client, &node_name).await;
                tokio::time::delay_for(sleep_interval).await;
            }
        });

        // This informer listens for pod events.
        let provider = self.provider.clone();
        let node_selector = format!("spec.nodeName={}", self.config.node_name);
        let pod_informer = tokio::task::spawn(async move {
            // Create our informer and start listening.
            let params = ListParams {
                field_selector: Some(node_selector),
                ..Default::default()
            };
            let informer = Informer::new(client, params, Resource::all::<KubePod>());
            loop {
                let mut stream = informer.poll().await.expect("informer poll failed").boxed();
                while let Some(event) = stream.try_next().await.unwrap() {
                    debug!("Handling Kubernetes pod event: {:?}", event);
                    match provider.lock().await.handle_event(event).await {
                        Ok(()) => debug!("Handled Kubernetes event successfully"),
                        Err(e) => warn!("Error handling pod event: {}", e),
                    };
                }
            }
        });

        // Start the webserver
        let webserver = start_webserver(self.provider.clone(), &self.config.server_config);

        // FIXME: If any of these dies, we should crash the Kubelet and let it restart.
        // A Future that will complete as soon as either spawned task fails
        let threads = async {
            futures::try_join!(node_updater, pod_informer)?;
            Ok(())
        };

        // Return an error as soon as either the webserver or the threads error
        futures::try_join!(webserver, threads)?;

        Ok(())
    }
}

// We cannot `#[derive(Clone)]` because that would place the
// unnecessary `P: Clone` constraint.
impl<P> Clone for Kubelet<P> {
    fn clone(&self) -> Self {
        Self {
            provider: self.provider.clone(),
            kube_config: self.kube_config.clone(),
            config: self.config.clone(),
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::Pod;
    use k8s_openapi::api::core::v1::{
        Container, EnvVar, EnvVarSource, ObjectFieldSelector, PodSpec, PodStatus,
    };
    use kube::api::ObjectMeta;
    use std::collections::BTreeMap;

    fn mock_client() -> kube::Client {
        kube::config::Configuration {
            base_path: ".".to_string(),
            client: reqwest::Client::new(),
            default_ns: " ".to_string(),
        }
        .into()
    }

    struct MockProvider;

    #[async_trait::async_trait]
    impl Provider for MockProvider {
        const ARCH: &'static str = "mock";
        async fn add(&self, _pod: Pod) -> anyhow::Result<()> {
            Ok(())
        }
        async fn modify(&self, _pod: Pod) -> anyhow::Result<()> {
            Ok(())
        }
        async fn delete(&self, _pod: Pod) -> anyhow::Result<()> {
            Ok(())
        }
        async fn logs(
            &self,
            _namespace: String,
            _pod: String,
            _container: String,
        ) -> anyhow::Result<Vec<u8>> {
            Ok(vec![])
        }
    }

    #[tokio::test]
    async fn test_env_vars() {
        let container = Container {
            env: Some(vec![
                EnvVar {
                    name: "first".into(),
                    value: Some("value".into()),
                    ..Default::default()
                },
                EnvVar {
                    name: "second".into(),
                    value_from: Some(EnvVarSource {
                        field_ref: Some(ObjectFieldSelector {
                            field_path: "metadata.labels.label".into(),
                            ..Default::default()
                        }),
                        ..Default::default()
                    }),
                    ..Default::default()
                },
                EnvVar {
                    name: "third".into(),
                    value_from: Some(EnvVarSource {
                        field_ref: Some(ObjectFieldSelector {
                            field_path: "metadata.annotations.annotation".into(),
                            ..Default::default()
                        }),
                        ..Default::default()
                    }),
                    ..Default::default()
                },
                EnvVar {
                    name: "NAME".into(),
                    value_from: Some(EnvVarSource {
                        field_ref: Some(ObjectFieldSelector {
                            field_path: "metadata.name".into(),
                            ..Default::default()
                        }),
                        ..Default::default()
                    }),
                    ..Default::default()
                },
                EnvVar {
                    name: "NAMESPACE".into(),
                    value_from: Some(EnvVarSource {
                        field_ref: Some(ObjectFieldSelector {
                            field_path: "metadata.namespace".into(),
                            ..Default::default()
                        }),
                        ..Default::default()
                    }),
                    ..Default::default()
                },
                EnvVar {
                    name: "HOST_IP".into(),
                    value_from: Some(EnvVarSource {
                        field_ref: Some(ObjectFieldSelector {
                            field_path: "status.hostIP".into(),
                            ..Default::default()
                        }),
                        ..Default::default()
                    }),
                    ..Default::default()
                },
                EnvVar {
                    name: "POD_IP".into(),
                    value_from: Some(EnvVarSource {
                        field_ref: Some(ObjectFieldSelector {
                            field_path: "status.podIP".into(),
                            ..Default::default()
                        }),
                        ..Default::default()
                    }),
                    ..Default::default()
                },
            ]),
            ..Default::default()
        };
        let name = "my-name".to_string();
        let namespace = Some("my-namespace".to_string());
        let mut labels = BTreeMap::new();
        labels.insert("label".to_string(), "value".to_string());
        let mut annotations = BTreeMap::new();
        annotations.insert("annotation".to_string(), "value".to_string());
        let pod = Pod::new(KubePod {
            metadata: Some(ObjectMeta {
                labels: Some(labels),
                annotations: Some(annotations),
                name: Some(name),
                namespace,
                ..Default::default()
            }),
            spec: Some(PodSpec {
                service_account_name: Some("svc".to_string()),
                ..Default::default()
            }),
            status: Some(PodStatus {
                host_ip: Some("10.21.77.1".to_string()),
                pod_ip: Some("10.21.77.2".to_string()),
                ..Default::default()
            }),
        });
        let env = MockProvider::env_vars(&container, &pod, &mock_client()).await;

        assert_eq!(
            "value",
            env.get("first").expect("key first should exist").as_str()
        );

        assert_eq!(
            "value",
            env.get("second").expect("metadata.labels.label").as_str()
        );
        assert_eq!(
            "value",
            env.get("third")
                .expect("metadata.annotations.annotation")
                .as_str()
        );
        assert_eq!("my-name", env.get("NAME").expect("metadata.name").as_str());
        assert_eq!(
            "my-namespace",
            env.get("NAMESPACE").expect("metadata.namespace").as_str()
        );
        assert_eq!("10.21.77.2", env.get("POD_IP").expect("pod_ip").as_str());
        assert_eq!("10.21.77.1", env.get("HOST_IP").expect("host_ip").as_str());
    }
}
