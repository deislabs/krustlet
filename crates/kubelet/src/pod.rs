use crate::Phase;
use k8s_openapi::api::core::v1::Container as KubeContainer;
use k8s_openapi::api::core::v1::Pod as KubePod;
use kube::{
    api::{Api, PatchParams},
    client::APIClient,
};
use log::{debug, error, info};

/// A Kubernetes Pod
///
/// This is a new type around the k8s_openapi Pod definition
/// providing convenient accessor methods
#[derive(Default, Debug)]
pub struct Pod(KubePod);

impl Pod {
    /// Construct a new Pod
    pub fn new(inner: KubePod) -> Pod {
        Pod(inner)
    }

    /// Get the name of the pod
    pub fn name(&self) -> &str {
        self.0
            .metadata
            .as_ref()
            .and_then(|m| m.name.as_deref())
            .expect("Pod name should always be set but was not")
    }

    /// Get the pod's namespace
    ///
    /// Returns "default" if no namespace was explictily set
    pub fn namespace(&self) -> &str {
        let metadata = self.0.metadata.as_ref();
        metadata
            .and_then(|m| m.namespace.as_deref())
            .unwrap_or("default")
    }

    /// Get the pod's node_selector map
    pub fn node_selector(&self) -> Option<&std::collections::BTreeMap<String, String>> {
        self.0.spec.as_ref()?.node_selector.as_ref()
    }

    /// Get the pod's service account name
    pub fn service_account_name(&self) -> Option<&str> {
        let spec = self.0.spec.as_ref()?;
        spec.service_account_name.as_deref()
    }

    /// Get the pod's host ip
    pub fn host_ip(&self) -> Option<&str> {
        let status = self.0.status.as_ref()?;
        status.host_ip.as_deref()
    }

    /// Get the pod's ip
    pub fn pod_ip(&self) -> Option<&str> {
        let status = self.0.status.as_ref()?;
        status.pod_ip.as_deref()
    }

    /// Get an iterator over the pod's labels
    pub fn labels(&self) -> &std::collections::BTreeMap<String, String> {
        self.0
            .metadata
            .as_ref()
            .and_then(|metadata| metadata.labels.as_ref())
            .unwrap_or_else(|| &EMPTY_MAP)
    }

    ///  Get the pod's annotations
    pub fn annotations(&self) -> &std::collections::BTreeMap<String, String> {
        self.0
            .metadata
            .as_ref()
            .and_then(|metadata| metadata.annotations.as_ref())
            .unwrap_or_else(|| &EMPTY_MAP)
    }

    ///  Get a specific annotation from the pod
    pub fn get_annotation(&self, key: &str) -> Option<&str> {
        Some(self.annotations().get(key)?.as_str())
    }

    /// Patch the pod status to update the phase.
    pub async fn patch_status(&self, client: APIClient, phase: &Phase) {
        let status = serde_json::json!(
            {
                "metadata": {
                    "resourceVersion": "",
                },
                "status": {
                    "phase": phase
                }
            }
        );

        let data = serde_json::to_vec(&status).expect("Should always serialize");
        let name = self.name();
        let api: Api<KubePod> = Api::namespaced(client, self.namespace());
        match api.patch_status(&name, &PatchParams::default(), data).await {
            Ok(o) => {
                info!("Pod status for {} set to {:?}", name, phase);
                debug!("Pod status returned: {:#?}", o.status)
            }
            Err(e) => error!("Pod status update failed for {}: {}", name, e),
        }
    }

    /// Get a pod's containers
    pub fn containers(&self) -> &Vec<KubeContainer> {
        self.0
            .spec
            .as_ref()
            .map(|s| &s.containers)
            .unwrap_or_else(|| &EMPTY_VEC)
    }

    /// Turn the Pod into the Kubernetes API version of a Pod
    pub fn into_kube_pod(self) -> KubePod {
        self.0
    }

    /// Turn a reference to a Pod into a reference to the Kubernetes API version of a Pod
    pub fn as_kube_pod(&self) -> &KubePod {
        &self.0
    }
}

impl std::convert::From<KubePod> for Pod {
    fn from(api_pod: KubePod) -> Self {
        Self(api_pod)
    }
}

impl<'a> std::convert::From<&'a Pod> for &'a KubePod {
    fn from(pod: &'a Pod) -> Self {
        &pod.0
    }
}
impl std::convert::From<Pod> for KubePod {
    fn from(pod: Pod) -> Self {
        pod.0
    }
}

lazy_static::lazy_static! {
    static ref EMPTY_MAP: std::collections::BTreeMap<String, String> = std::collections::BTreeMap::new();
    static ref EMPTY_VEC: Vec<KubeContainer> = Vec::new();
}
