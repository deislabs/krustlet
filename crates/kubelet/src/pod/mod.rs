//! `pod` is a collection of utilities surrounding the Kubernetes pod API.
mod handle;
mod queue;
mod status;

pub use handle::{key_from_pod, pod_key, Handle};
pub(crate) use queue::Queue;
pub use status::{update_status, Phase, Status};

use std::collections::HashMap;

use chrono::{DateTime, Utc};
use k8s_openapi::api::core::v1::{
    Container as KubeContainer, ContainerStatus as KubeContainerStatus, Pod as KubePod,
    Volume as KubeVolume,
};
use kube::api::{Api, Meta, PatchParams};
use log::{debug, error};

/// A Kubernetes Pod
///
/// This is a new type around the k8s_openapi Pod definition
/// providing convenient accessor methods
#[derive(Default, Debug, Clone)]
pub struct Pod(KubePod);

impl Pod {
    /// Construct a new Pod
    pub fn new(inner: KubePod) -> Self {
        Self(inner)
    }

    /// Get the name of the pod
    pub fn name(&self) -> &str {
        self.0
            .metadata
            .as_ref()
            .and_then(|m| m.name.as_deref())
            .expect("Pod name should always be set but was not")
    }

    /// Get the pod's namespace
    ///
    /// Returns "default" if no namespace was explictily set
    pub fn namespace(&self) -> &str {
        let metadata = self.0.metadata.as_ref();
        metadata
            .and_then(|m| m.namespace.as_deref())
            .unwrap_or("default")
    }

    /// Get the pod's node_selector map
    pub fn node_selector(&self) -> Option<&std::collections::BTreeMap<String, String>> {
        self.0.spec.as_ref()?.node_selector.as_ref()
    }

    /// Get the pod's service account name
    pub fn service_account_name(&self) -> Option<&str> {
        let spec = self.0.spec.as_ref()?;
        spec.service_account_name.as_deref()
    }

    /// Get the pod volumes
    pub fn volumes(&self) -> Option<&Vec<KubeVolume>> {
        let spec = self.0.spec.as_ref()?;
        spec.volumes.as_ref()
    }

    /// Get the pod's host ip
    pub fn host_ip(&self) -> Option<&str> {
        let status = self.0.status.as_ref()?;
        status.host_ip.as_deref()
    }

    /// Get the pod's ip
    pub fn pod_ip(&self) -> Option<&str> {
        let status = self.0.status.as_ref()?;
        status.pod_ip.as_deref()
    }

    /// Get an iterator over the pod's labels
    pub fn labels(&self) -> &std::collections::BTreeMap<String, String> {
        self.0.meta().labels.as_ref().unwrap_or_else(|| &EMPTY_MAP)
    }

    ///  Get the pod's annotations
    pub fn annotations(&self) -> &std::collections::BTreeMap<String, String> {
        self.0
            .meta()
            .annotations
            .as_ref()
            .unwrap_or_else(|| &EMPTY_MAP)
    }

    /// Indicate if this pod is a static pod.
    /// TODO: A missing owner_references field was an indication of static pod in my testing but I
    /// dont know how reliable this is.
    pub fn is_static(&self) -> bool {
        self.0.meta().owner_references.is_none()
    }

    /// Indicate if this pod is part of a Daemonset
    pub fn is_daemonset(&self) -> bool {
        if let Some(owners) = &self.0.meta().owner_references {
            for owner in owners {
                if owner.kind == "DaemonSet" {
                    return true;
                }
            }
        }
        false
    }

    ///  Get a specific annotation from the pod
    pub fn get_annotation(&self, key: &str) -> Option<&str> {
        Some(self.annotations().get(key)?.as_str())
    }

    /// Get the deletionTimestamp if it exists
    pub fn deletion_timestamp(&self) -> Option<&DateTime<Utc>> {
        self.0.meta().deletion_timestamp.as_ref().map(|t| &t.0)
    }

    /// Patch the pod status using the given status information.
    pub async fn patch_status(&self, client: kube::Client, status: Status) {
        let name = self.name();
        let api: Api<KubePod> = Api::namespaced(client, self.namespace());
        let current_status = match api.get(name).await {
            Ok(p) => match p.status {
                Some(s) => s,
                None => {
                    error!("Pod is missing status information. This should not occur");
                    return;
                }
            },
            Err(e) => {
                error!("Unable to fetch current status of pod {}, aborting status patch (will be retried on next status update): {:?}", name, e);
                return;
            }
        };

        // This section figures out what the current phase of the pod should be
        // based on the container statuses
        let current_statuses = status
            .container_statuses
            .into_iter()
            .map(|s| (s.0.clone(), s.1.to_kubernetes(s.0)))
            .collect::<HashMap<String, KubeContainerStatus>>();
        // Filter out any ones we are updating and then combine them all together
        let mut container_statuses = current_status
            .container_statuses
            .unwrap_or_default()
            .into_iter()
            .filter(|s| !current_statuses.contains_key(&s.name))
            .collect::<Vec<KubeContainerStatus>>();
        container_statuses.extend(current_statuses.into_iter().map(|(_, v)| v));
        let mut num_succeeded: usize = 0;
        let mut failed = false;
        // TODO(thomastaylor312): Add inferring a message from these container
        // statuses if there is no message passed in the Status object
        for status in container_statuses.iter() {
            // Basically anything is considered running phase in kubernetes
            // unless it is explicitly exited, so don't worry about considering
            // that state. We only really need to check terminated
            if let Some(terminated) = &status.state.as_ref().unwrap().terminated {
                if terminated.exit_code != 0 {
                    failed = true;
                    break;
                } else {
                    num_succeeded += 1
                }
            }
        }
        // is there ever a case when we get a status that we should end up in Phase unknown?
        let phase = if num_succeeded == container_statuses.len() {
            Phase::Succeeded
        } else if failed {
            Phase::Failed
        } else {
            Phase::Running
        };

        let json_status = serde_json::json!(
            {
                "metadata": {
                    "resourceVersion": "",
                },
                "status": {
                    "phase": phase,
                    "message": status.message,
                    "containerStatuses": container_statuses
                }
            }
        );

        debug!("Setting pod status for {} using {:?}", name, json_status);

        let data = serde_json::to_vec(&json_status).expect("Should always serialize");
        match api.patch_status(&name, &PatchParams::default(), data).await {
            Ok(o) => debug!("Pod status returned: {:#?}", o.status),
            Err(e) => error!("Pod status update failed for {}: {}", name, e),
        }
    }

    /// Get a pod's containers
    pub fn containers(&self) -> &Vec<KubeContainer> {
        self.0
            .spec
            .as_ref()
            .map(|s| &s.containers)
            .unwrap_or_else(|| &EMPTY_VEC)
    }

    /// Turn the Pod into the Kubernetes API version of a Pod
    pub fn into_kube_pod(self) -> KubePod {
        self.0
    }

    /// Turn a reference to a Pod into a reference to the Kubernetes API version of a Pod
    pub fn as_kube_pod(&self) -> &KubePod {
        &self.0
    }
}

impl std::convert::From<KubePod> for Pod {
    fn from(api_pod: KubePod) -> Self {
        Self(api_pod)
    }
}

impl<'a> std::convert::From<&'a Pod> for &'a KubePod {
    fn from(pod: &'a Pod) -> Self {
        &pod.0
    }
}
impl std::convert::From<Pod> for KubePod {
    fn from(pod: Pod) -> Self {
        pod.0
    }
}

lazy_static::lazy_static! {
    static ref EMPTY_MAP: std::collections::BTreeMap<String, String> = std::collections::BTreeMap::new();
    static ref EMPTY_VEC: Vec<KubeContainer> = Vec::new();
}
