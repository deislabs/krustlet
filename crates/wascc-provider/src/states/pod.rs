pub(crate) mod crash_loop_backoff;
pub(crate) mod error;
pub(crate) mod image_pull;
pub(crate) mod image_pull_backoff;
pub(crate) mod registered;
pub(crate) mod running;
pub(crate) mod starting;
pub(crate) mod terminated;
pub(crate) mod volume_mount;
