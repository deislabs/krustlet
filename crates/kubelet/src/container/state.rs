//! Functions for running Container state machines.
use crate::container::{patch_container_status, Status};
use crate::container::{Container, ContainerKey};
use crate::pod::Pod;
use crate::state::{ResourceState, State, Transition};
use k8s_openapi::api::core::v1::Pod as KubePod;
use kube::api::Api;
use log::{debug, error, warn};
use std::sync::Arc;
use tokio::sync::RwLock;

/// Iteratively evaluate state machine until it returns Complete.
pub async fn run_to_completion<S: ResourceState<Manifest = Container> + Send + Sync + 'static>(
    client: &kube::Client,
    initial_state: impl State<S, Status>,
    container_state: &mut S,
    pod: Arc<RwLock<Pod>>,
    container_name: ContainerKey,
) {
    let (pod_name, api) = {
        let initial_pod = pod.read().await.clone();
        let namespace = initial_pod.namespace().to_string();
        let name = initial_pod.name().to_string();
        let api: Api<KubePod> = Api::namespaced(client.clone(), &namespace);
        (name, api)
    };

    let mut state: Box<dyn State<S, Status>> = Box::new(initial_state);

    loop {
        debug!(
            "Pod {} container {} entering state {:?}",
            &pod_name, container_name, state
        );

        let latest_pod = { pod.read().await.clone() };
        let latest_container = latest_pod.find_container(&container_name).unwrap();

        // TODO Make json_status -> status
        match state.json_status(container_state, &latest_container).await {
            Ok(status) => patch_container_status(&api, &latest_pod, &container_name, &status)
                .await
                .unwrap(),
            Err(e) => {
                warn!(
                    "Pod {} container {} status patch returned error: {:?}",
                    &pod_name, container_name, e
                );
            }
        }

        debug!(
            "Pod {} container {} executing state handler {:?}",
            &pod_name, container_name, state
        );
        let transition = { state.next(container_state, &latest_container).await };

        state = match transition {
            Transition::Next(s) => {
                debug!(
                    "Pod {} container {} transitioning to {:?}.",
                    &pod_name, container_name, s.state
                );
                s.state
            }
            Transition::Complete(result) => match result {
                Ok(()) => {
                    debug!(
                        "Pod {} container {} state machine exited without error",
                        &pod_name, container_name
                    );
                    // TODO Patch container success.
                    break;
                }
                Err(e) => {
                    error!(
                        "Pod {} container {} state machine exited with error: {:?}",
                        &pod_name, container_name, e
                    );
                    // TODO Patch container failure.
                    break;
                }
            },
        };
    }
}
