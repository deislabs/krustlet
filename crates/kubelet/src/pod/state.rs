//! Functions for running Pod state machines.
use crate::pod::{initialize_pod_container_statuses, Phase, Pod};
use crate::state::{ResourceState, State, Transition};
use k8s_openapi::api::core::v1::Pod as KubePod;
use kube::api::{Api, PatchParams};
use log::{debug, error, warn};
use std::sync::Arc;
use tokio::sync::RwLock;

async fn patch_status(api: &Api<KubePod>, name: &str, patch: serde_json::Value) {
    match serde_json::to_vec(&patch) {
        Ok(data) => match api.patch_status(&name, &PatchParams::default(), data).await {
            Ok(_) => (),
            Err(e) => {
                warn!("Pod {} error patching status: {:?}", name, e);
            }
        },
        Err(e) => {
            warn!(
                "Pod {} error serializing status patch {:?}: {:?}",
                name, &patch, e
            );
        }
    }
}

/// Iteratively evaluate state machine until it returns Complete.
pub async fn run_to_completion<PodState: ResourceState<Manifest = Pod> + Send + Sync + 'static>(
    client: &kube::Client,
    state: impl State<PodState>,
    pod_state: &mut PodState,
    pod: Arc<RwLock<Pod>>,
) {
    let (name, api) = {
        let initial_pod = pod.read().await.clone();
        let namespace = initial_pod.namespace().to_string();
        let name = initial_pod.name().to_string();
        let api: Api<KubePod> = Api::namespaced(client.clone(), &namespace);
        (name, api)
    };

    if initialize_pod_container_statuses(&name, Arc::clone(&pod), &api)
        .await
        .is_err()
    {
        return;
    }

    let mut state: Box<dyn State<PodState>> = Box::new(state);

    loop {
        debug!("Pod {} entering state {:?}", &name, state);

        let latest_pod = { pod.read().await.clone() };

        match state.json_status(pod_state, &latest_pod).await {
            Ok(patch) => {
                patch_status(&api, &name, patch).await;
            }
            Err(e) => {
                warn!("Pod {} status patch returned error: {:?}", &name, e);
            }
        }

        debug!("Pod {} executing state handler {:?}", &name, state);
        let transition = { state.next(pod_state, &latest_pod).await };

        state = match transition {
            Transition::Next(s) => {
                debug!("Pod {} transitioning to {:?}.", &name, s.state);
                s.state
            }
            Transition::Complete(result) => match result {
                Ok(()) => {
                    debug!("Pod {} state machine exited without error", &name);
                    break;
                }
                Err(e) => {
                    error!("Pod {} state machine exited with error: {:?}", &name, e);
                    let patch = serde_json::json!(
                        {
                            "metadata": {
                                "resourceVersion": "",
                            },
                            "status": {
                                "phase": Phase::Failed,
                                "reason": format!("{:?}", e),
                            }
                        }
                    );
                    patch_status(&api, &name, patch).await;
                    break;
                }
            },
        };
    }
}

#[derive(Default, Debug)]
/// Stub state machine for testing.
pub struct Stub;

#[async_trait::async_trait]
impl<P: 'static + Sync + Send + ResourceState<Manifest = Pod>> State<P> for Stub {
    async fn next(self: Box<Self>, _state: &mut P, _manifest: &Pod) -> Transition<P> {
        Transition::Complete(Ok(()))
    }

    async fn json_status(
        &self,
        _pod_state: &mut P,
        _pod: &Pod,
    ) -> anyhow::Result<serde_json::Value> {
        Ok(serde_json::json!(null))
    }
}

#[cfg(test)]
mod test {
    use crate::pod::Pod;
    use crate::state::{ResourceState, State, Transition, TransitionTo};

    #[derive(Debug)]
    struct PodState;

    impl ResourceState for PodState {
        type Manifest = Pod;
    }

    #[derive(Debug)]
    struct ValidState;

    #[async_trait::async_trait]
    impl State<PodState> for ValidState {
        async fn next(
            self: Box<Self>,
            _pod_state: &mut PodState,
            _pod: &Pod,
        ) -> Transition<PodState> {
            Transition::Complete(Ok(()))
        }

        async fn json_status(
            &self,
            _pod_state: &mut PodState,
            _pod: &Pod,
        ) -> anyhow::Result<serde_json::Value> {
            Ok(serde_json::json!(null))
        }
    }

    #[test]
    fn it_can_transition_to_valid_state() {
        #[derive(Debug)]
        struct TestState;

        impl TransitionTo<ValidState> for TestState {}

        #[async_trait::async_trait]
        impl State<PodState> for TestState {
            async fn next(
                self: Box<Self>,
                _pod_state: &mut PodState,
                _pod: &Pod,
            ) -> Transition<PodState> {
                Transition::next(self, ValidState)
            }

            async fn json_status(
                &self,
                _pod_state: &mut PodState,
                _pod: &Pod,
            ) -> anyhow::Result<serde_json::Value> {
                Ok(serde_json::json!(null))
            }
        }
    }
}
