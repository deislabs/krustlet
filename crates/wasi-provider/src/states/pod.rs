use crate::ModuleRunContext;
use crate::ProviderState;
use async_trait::async_trait;
use kubelet::backoff::BackoffStrategy;
use kubelet::backoff::ExponentialBackoffStrategy;
use kubelet::pod::Pod;
use kubelet::pod::PodKey;
use kubelet::pod::Status;
use kubelet::state::common::{BackoffSequence, GenericPodState, ThresholdTrigger};
use kubelet::state::ResourceState;
use std::collections::HashMap;
use tokio::sync::mpsc;

pub(crate) mod completed;
pub(crate) mod initializing;
pub(crate) mod running;
pub(crate) mod starting;

/// State that is shared between pod state handlers.
pub struct PodState {
    key: PodKey,
    run_context: ModuleRunContext,
    errors: usize,
    image_pull_backoff_strategy: ExponentialBackoffStrategy,
    crash_loop_backoff_strategy: ExponentialBackoffStrategy,
}

#[async_trait]
impl ResourceState for PodState {
    type Manifest = Pod;
    type Status = Status;
    type SharedState = ProviderState;
    async fn async_drop(self, provider_state: &mut Self::SharedState) {
        {
            let mut handles = provider_state.handles.write().await;
            handles.remove(&self.key);
        }
    }
}

impl PodState {
    pub fn new(pod: &Pod) -> Self {
        let (tx, rx) = mpsc::channel(pod.all_containers().len());
        let run_context = ModuleRunContext {
            modules: Default::default(),
            volumes: Default::default(),
            status_sender: tx,
            status_recv: rx,
        };
        let key = PodKey::from(pod);
        PodState {
            key,
            run_context,
            errors: 0,
            image_pull_backoff_strategy: ExponentialBackoffStrategy::default(),
            crash_loop_backoff_strategy: ExponentialBackoffStrategy::default(),
        }
    }
}

#[async_trait]
impl GenericPodState for PodState {
    fn set_modules(&mut self, modules: HashMap<String, Vec<u8>>) {
        self.run_context.modules = modules;
    }
    fn set_volumes(&mut self, volumes: HashMap<String, kubelet::volume::Ref>) {
        self.run_context.volumes = volumes;
    }
    async fn backoff(&mut self, sequence: BackoffSequence) {
        let backoff_strategy = match sequence {
            BackoffSequence::ImagePull => &mut self.image_pull_backoff_strategy,
            BackoffSequence::CrashLoop => &mut self.crash_loop_backoff_strategy,
        };
        backoff_strategy.wait().await;
    }
    fn reset_backoff(&mut self, sequence: BackoffSequence) {
        let backoff_strategy = match sequence {
            BackoffSequence::ImagePull => &mut self.image_pull_backoff_strategy,
            BackoffSequence::CrashLoop => &mut self.crash_loop_backoff_strategy,
        };
        backoff_strategy.reset();
    }
    fn record_error(&mut self) -> ThresholdTrigger {
        self.errors += 1;
        if self.errors > 3 {
            self.errors = 0;
            ThresholdTrigger::Triggered
        } else {
            ThresholdTrigger::Untriggered
        }
    }
}
