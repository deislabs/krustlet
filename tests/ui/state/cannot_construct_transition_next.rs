// Test that State<T> can only transition to State<T>
// edition:2018
extern crate async_trait;
extern crate kubelet;

use kubelet::state::{Transition, StateHolder, ResourceState};
use kubelet::pod::{Pod, Status, state::Stub};

struct PodState;
struct ProviderState;

#[async_trait::async_trait]
impl ResourceState for PodState {
    type Manifest = Pod;
    type Status = Status;
    type SharedState = ProviderState;
    async fn async_drop(self, _provider_state: &mut ProviderState) { }
}

fn main() {
    // This fails because `state` is a private field. Use Transition::next classmethod instead.
    let _transition = Transition::<PodState>::Next(StateHolder {
        state: Box::new(Stub),
    });
}
