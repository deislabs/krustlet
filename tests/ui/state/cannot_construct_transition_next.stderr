error[E0451]: field `state` of struct `StateHolder` is private
  --> $DIR/cannot_construct_transition_next.rs:23:9
   |
23 |         state: Box::new(Stub),
   |         ^^^^^^^^^^^^^^^^^^^^^ private field

error: aborting due to previous error

For more information about this error, try `rustc --explain E0451`.
